Rework/configuration

tchapi opened the pull request:
See https://github.com/sabre-io/Baikal/issues/831

DeepDiver1975 commented on Core/Frameworks/Baikal/Core/Server.php:
```diff
@@ -131,6 +132,10 @@ function start() {
 
+        try {
+            $config = Yaml::parseFile(PROJECT_PATH_CONFIG . "config.yaml");
+        } catch (\Exception $e) {}
```
Log the exception?

tchapi replied:
Sure — do you mean `error_log()` for instance, or is there another helper that I should use ?

tchapi replied:
_see my last commit_

DeepDiver1975 commented on Core/Frameworks/Baikal/Model/Config.php:
```diff
@@ -27,18 +27,25 @@
+            $config = Yaml::parseFile($this->sConfigFilePath);
+            $aConfig = $config['parameters'];
+        } catch(\Exception $e) {
+            $aConfig = static::getDefaultConfig();
```
Log Here as well?

DeepDiver1975 commented on Core/Frameworks/Baikal/Model/Config/Standard.php:
```diff
@@ -110,7 +113,11 @@ function formMorphologyForThisModelInstance() {
-        if (!defined("BAIKAL_ADMIN_PASSWORDHASH") || trim(BAIKAL_ADMIN_PASSWORDHASH) === "") {
+        try {
+            $config = Yaml::parseFile(PROJECT_PATH_CONFIG . "system.yaml");
+        } catch(\Exception $e) {}
```
Log?

DeepDiver1975 commented on Core/Frameworks/Baikal/Model/User.php:
```diff
@@ -279,6 +281,11 @@ static function bigicon() {
+
+        try {
+            $config = Yaml::parseFile(PROJECT_PATH_CONFIG . "config.yaml");
+        } catch(\Exception $e) {}
```
Log?

tchapi commented:
I notice the failed builds too, I will try to correct when I can

ByteHamster commented:
While we already touch the configuration, I think it could be simplified to one single file only. That one file could have sections that represent the previous two files. Additionally, I think that the database configuration could be its own section. Maybe we could structure the file in the following sections:

- System: version, encryption, admin password
- Database 
- Server: cal enabled, dav enabled, invite mail

Not sure about the naming, though.

---

Most projects seem to use lower case yaml configuration keys. I don't want to touch the config multiple times, so it would be great if you could directly make it standard.

tchapi commented:
Understood and agree for the lowercase, I can change that.

For the single file with the different sections though, I tried and dropped the idea, because that would mean we need to change the whole framework with it, since having two files is deeply ingrained in the way it works (_and the way the install UI and the settings UI are built_). Considering my knowledge of the framework, I would likely break more things that solve anything, to be honest.

evert commented:
Looks great =) I'll leave it up to @ByteHamster to fully approve and merge and stuff.

Random thought, but doesn't have to be part of this PR... it would be so nice if the baikal version can move from the config file to the database. Would solve a lot of upgrade problems for people I think. Since upgrades is mostly database-patch based, it would make sense that the database version is also tied to the db itself.

Speaking of upgrades: how will that work?

ByteHamster commented:
> For the single file with the different sections though, I tried and dropped the idea, because that would mean we need to change the whole framework with it, since having two files is deeply ingrained in the way it works (and the way the install UI and the settings UI are built). 

I still think that if we already touch the config, we should do it properly. Wouldn't it be possible to do something like the following in `Config::persist`?

```php
$config = Yaml::parseFile($this->sConfigFilePath);
$config["system"] = $this->aData;
file_put_contents($this->sConfigFilePath, Yaml::dump($config));
```

That would allow to have one file with two sections. Every time one of the files is written, we just read it and update the corresponding section.

tchapi commented:
Hello @ByteHamster 

As I said in https://github.com/sabre-io/Baikal/issues/831, it's _really_ hard to work with the actual code structure and framework, so it doesn't just boil down to changing `Config:persist` in fact. If I'm touching this, I'm basically going down the rabbit hole of refactoring various other places, which I'm not enclined to do, since I'd rather start anew on a robust framework.

The double config file paradigm is not ideal I agree, but it's not a problem either I think.

@evert for the futur upgrades, I guess nothing changes : the next update that needs to change or add a setting will need to have a section in the `VersionUpgrade.php` file to account for it.

@DeepDiver1975 are my changes ok ?

Best regards

ByteHamster commented:
If I try to upgrade my existing installation to your branch, I get the following message on the `Baïkal Settings` page. When pressing `Save changes`, the screen refreshes but does not go to the next step.

```
Notice: Undefined index: project_timezone in baikal/Core/Frameworks/Baikal/Framework.php on line 64
Notice: date_default_timezone_set(): Timezone ID '' is invalid in htdocs/baikal/Core/Frameworks/Baikal/Framework.php on line 64
```

Even when deleting the `config` and `Specific` folder content, I can not set up a fresh install.

Looks like `$this->oForm->submitted()` never returns `true` in `Core/Frameworks/BaikalAdmin/Controller/Install/Initialize.php`. I noticed that my browser first does a request to `baikal/html/admin/install/?/database` with the form parameters, then gets forwarded to the same page again (`302 Found`). The second request no longer contains the POST headers.

![image](https://user-images.githubusercontent.com/5811634/67633454-760ce580-f8b0-11e9-8e2a-9c2867dd8dac.png)
![image](https://user-images.githubusercontent.com/5811634/67633456-7c9b5d00-f8b0-11e9-8a06-10930d6bb884.png)

Do you have an idea why this could happen? On `master`, the setup works fine.

tchapi commented:
Hello

It seems to be linked with the lowercase / uppercase change. I've overlooked some files when making the change, I've fixed it now. There may be more, though, I'm still looking.

ByteHamster commented:
Just tried your PR again.

- Users have to go through the installer again. It would be great if there was a little message telling the users that their events are safe. Something like:
  > Our configuration format was updated. You need to go through the installer again but we pre-filled most values with the ones from the old installation. The database with your contacts and events stays untouched. A backup is still advisable.
- Users have to enter their password again. What about taking the chance to update to a better hashing algorithm? We currently use md5. Suggestion: Change `getPasswordHashForPassword` to
  ```
  hash("sha256", $this->get("username") . ':'
    . $config['parameters']['baikal_auth_realm'] . ':' . $sPassword)
  ```
- During upgrade, I get the following error. The content of the `SQLite file path` text field is `/mnt/.../baikal/Specific/db/db.sqlite`.
    ```
    ParseError: syntax error, unexpected '/', expecting ';' in baikal/Core/Frameworks/BaikalAdmin/Controller/Install/Database.php(154) : eval()'d code:1
    Stack trace:
    #0 [internal function]: BaikalAdmin\Controller\Install\Database->validateConnection(Object(Formal\Form), Object(Formal\Form\Morphology))
    #1 baikal/Core/Frameworks/Formal/Form.php(202): call_user_func(Array, Object(Formal\Form), Object(Formal\Form\Morphology))
    #2 baikal/Core/Frameworks/BaikalAdmin/Controller/Install/Database.php(59): Formal\Form->execute()
    #3 baikal/Core/Frameworks/Flake/Core/Render/Container.php(62): BaikalAdmin\Controller\Install\Database->execute()
    #4 baikal/Core/Frameworks/Flake/Controller/Page.php(83): Flake\Core\Render\Container->execute()
    #5 baikal/Core/Frameworks/BaikalAdmin/WWWRoot/install/index.php(107): Flake\Controller\Page->render()
    #6 {main}
    ```
- I still think that it should be possible to switch to one single config file. I can give it a try after this PR is ready. Switching to a single config file would make it possible to have a clean start for the rewritten admin interface.

tchapi commented:
Hi,

1. Added a little message for that
2. Changed the hashing algorithm — but I have a doubt — you're talking about the admin user (that has to enter its password again that's true) or all the users (that don't have to do so - and that's the piece of code you linked) ? I changed both. But I'm not sure now, re-reading your comment, that it's what you wanted though.
3. Fixed the bug (well, removed the _horrible_ `eval` because I really don't understand its purpose)
4. On this single-config file thing, be my guest :) I have created my own server based on sabre/dav (it's on my github if interested) so after this PR, I'm not sure I'll have a lot of time to dedicate to Baïkal. I will continue to try to help on [sabre/dav](https://github.com/sabre-io/dav) though (I've already submitted a few PRs)

BR

ByteHamster commented:
> Changed the hashing algorithm — but I have a doubt — you're talking about the admin user (that has to enter its password again that's true) or all the users (that don't have to do so - and that's the piece of code you linked) ? I changed both. But I'm not sure now, re-reading your comment, that it's what you wanted though.

Oops, you are right. I talked about the admin interface. We don't re-enter the user passwords, so their hash should not be changed.

> I have created my own server based on sabre/dav (it's on my github if interested) so after this PR, I'm not sure I'll have a lot of time to dedicate to Baïkal.

Woah, that was fast. Looks nice. An advantage of updating Baikal instead of starting your own project would be that it has an existing user base. Maybe we can think about merging the two projects later.

> I will continue to try to help on sabre/dav though (I've already submitted a few PRs)

Great to hear. So in a way, you still help developing Baikal ;)

tchapi commented:
I reverted https://github.com/sabre-io/Baikal/pull/835/commits/b28261fa0e6f97723c32cfc2586f6d7e1f91739f and changed the algorithm only for the admin interface :)

phil-davis commented:
@tchapi it would be useful to rebase this to latest `master` to verify that all passes with the latest CI.

@ByteHamster @DeepDiver1975 or whoever - are there code review comments still to be addressed or?

ByteHamster commented:
I want to look at the PR and try to replace the two config files with a single one. @tchapi said that it requires a lot of refactoring. I want to experiment a bit before believing this :D

tchapi commented:
Be my guest — I'll leave the rebase to you since it's been more than 3 months now ...

ByteHamster commented:
@tchapi Thanks for working on this. I created PR #899 as a replacement to continue your work. Using my own branch is easier than having to push to yours.
